repositories/repositories.bzl

```diff
@@ -79,26 +79,41 @@ def repositories():
             name = "containerregistry",
             sha256 = "a8cdf2452323e0fefa4edb01c08b2ec438c9fa3192bc9f408b89287598c12abc",
             strip_prefix = "containerregistry-" + CONTAINERREGISTRY_RELEASE[1:],
+            # containerregistry assumes a repository named @httplib2, but
+            # https://github.com/bazelbuild/bazel/issues/3998 means that this
+            # causes problems with the import path. Instead, we use
+            # repo_mapping and a fully-qualified repository name.
+            repo_mapping = {"@httplib2": "@com_github_httplib2_httplib2"},
             urls = [("https://github.com/google/containerregistry/archive/" +
                      CONTAINERREGISTRY_RELEASE + ".tar.gz")],
         )
 
     # TODO(mattmoor): Remove all of this (copied from google/containerregistry)
     # once transitive workspace instantiation lands.
 
-    if "httplib2" not in excludes:
+    if "com_github_httplib2_httplib2" not in excludes:
         # TODO(mattmoor): Is there a clean way to override?
         http_archive(
-            name = "httplib2",
+            name = "com_github_httplib2_httplib2",
             build_file_content = """
 py_library(
    name = "httplib2",
-   srcs = glob(["**/*.py"]),
-   data = ["cacerts.txt"],
+   srcs = select({
```

Comment:
If your issue is related to building this library in the host configuration, then these selects won't do much since they'll always match whichever value you pass to `--host_force_python`. Select reads the version from the configuration, but the whole problem with the host config is that it doesn't allow you to use any version besides whatever is passed into that flag.

Reply:
> the whole problem with the host config is that it doesn't allow you to use any version besides whatever is passed into that flag.

Hmm, I'm not sure I understand. I want the build to use the source code that matches how it will be executed, ie with --host_force_python=PY<x>, it should use the python<x> source.

My understanding: currently, genrules execute under PY2 regardless of --host_force_python (at least, --host_force_python=PY3 doesn't seem to break rules_docker). With --incompatible_use_python_toolchains, genrules execute under PY3 by default, which breaks rules_docker unless --host_force_python=PY2 is used. By adding select()s, httplib2 uses the source that matches the value of --host_force_python, so genrules work with either Python version. Does that make sense?

What I'm (even) less sure about is how this select() affects the version of Python used when building in the non-host configuration: would this change make it impossible to have a PY2 py_binary that depends on httplib2 (assuming --host_force_python=PY3)? Is it possible to have a py_library that uses different source depending on whether it is depended upon by a PY2 or PY3 py_binary?

Reply:
Genrule tools (and other things in host configuration) will be configured with the python version set to whatever `--host_force_python` is, which defaults to `PY3` (at least when `--incompatible_py3_is_default` is enabled, which is flipped to true in 0.24). The `python_version` attribute is totally ignored for host-configured things.

However, if Python toolchains are not enabled (you haven't set `--incompatible_use_python_toolchains`), it's possible that these tools are running with a Python 2 interpreter at execution time even though they're configured to use Python 3 due to the host configuration (see bazelbuild/bazel#4815). So the limitation of what interpreter you can request in the host config, is canceled out by a bug that doesn't give you the interpreter you requested. This may be why setting `--host_force_python=PY3` didn't break rules_docker. Enabling toolchains fixes the latter bug and reveals the need to set `--host_force_python` for the build to work.

bazelbuild/bazel#4815 never affected how selects resolve. In the host config, they'll always think the version is whatever `--host_force_python` is, regardless of whether or not you're using toolchains -- and that's correct in the sense that that *is* what the version is in the host config.

Outside of the host config, you can use these selects in `py_library` and elsewhere to vary the behavior depending on whether it's underneath a PY2 or PY3 `py_binary`.